### PR TITLE
⭐️New: Add rule `no-template-shadow`.

### LINKED ISSUE #101: Rule Proposal: `vue/no-shadow`
## Please describe what the rule should do:

`no-shadow` should report variable definitions of `v-for` directives or `scope` attributes if those shadows the variables in parent scopes.

## What category of rule is this? (place an "X" next to just one item)

[ ] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

## Provide 2-3 code examples that this rule will warn about:

```html
<template>
    <ol v-for="i in 5">
        <ol v-for="i in 5"><!-- "i" is already declared in the upper scope. -->
            <li>item</li>
        </ol>
    </ol>
</template>
<script>
export default {
}
</script>
```

### PR DESCRIPTION
**Pull request by armano2**
fixes #101

### CODE REVIEW

File: tests/lib/rules/no-template-shadow.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+      </script>`,
+      errors: [{
+        message: "Variable 'i' is already declared in the upper scope.",
+        type: 'Identifier'
+      }, {
+        message: "Variable 'i' is already declared in the upper scope.",
+        type: 'Identifier'
+      }]
+    },
+    {
```
Comment (Aladdin-ADD) [inline tests/lib/rules/no-template-shadow.js]: I like to see some tests like this:
```html
<template>
<ul v-for="i in 5">
  <li>
    <span v-for="j in 10">{{i}},{{j}}</span>
  </li>
</ul>
</template>
```

File: docs/rules/no-template-shadow.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+`no-shadow` should report variable definitions of v-for directives or scope attributes if those shadows the variables in parent scopes.
+
+## :book: Rule Details
+
+This rule aims to eliminate shadowed variable declarations of v-for directives or scope attributes.
+
+:-1: Examples of **incorrect** code for this rule:
+
+```html
```
Comment (michalsnik) [inline docs/rules/no-template-shadow.md]: I think we should present two or three cases here. First - a simple template example, and second with shadowed `data` or `method` property declarations. e.g.:
1.
```
<template>
   <div>
     <div v-for="i in 5">
       <div v-for="i in 10"></div>
     </div>
   </div>
 </template>
```

2.
```
<template>
   <div>
     <div v-for="i in 5"></div>
   </div>
 </template>
<script>
  export default {
    data () {
      return {
        i: 10
      }
    }
  }
</script>
```

File: lib/rules/no-template-shadow.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+const GROUP_NAMES = ['props', 'computed', 'data', 'methods']
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Disallow variable declarations from shadowing variables declared in the outer scope.',
+      category: 'Possible Errors',
```
Comment (michalsnik) [inline lib/rules/no-template-shadow.js]: I think it should belong to `essential` group

File: lib/utils/index.js [hunk 4]
```diff
@@ -26,7 +26,7 @@ module.exports = {
    *
    * @param {RuleContext} context The rule context to use parser services.
    * @param {Object} templateBodyVisitor The visitor to traverse the template body.
-   * @param {Object} scriptVisitor The visitor to traverse the script.
+   * @param {Object} [scriptVisitor] The visitor to traverse the script.
```
Comment (michalsnik) [inline lib/utils/index.js]: Why square brackets here?

File: tests/lib/rules/no-template-shadow.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    '',
+    '<template><div></div></template>',
+    '<template><div v-for="i in 5"></div></template>',
+    '<template><div v-for="i in 5"><div v-for="b in 5"></div></div></template>',
+    '<template><div v-for="i in 5"></div><div v-for="i in 5"></div></template>',
+    '<template> <ul v-for="i in 5"> <li> <span v-for="j in 10">{{i}},{{j}}</span> </li> </ul> </template>',
+    {
+      filename: 'test.vue',
+      code: `<template>
+        <div v-for="i in 5"></div>
```
Comment (michalsnik) [inline tests/lib/rules/no-template-shadow.js]: Let's add case with `i in f`, wdyt?

File: tests/lib/rules/no-template-shadow.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+      </template>
+      <script>
+        export default {
+          data: {
+            i: 7
+          }
+        }
+      </script>`,
+      errors: [{
+        message: "Variable 'i' is already declared in the upper scope.",
```
Comment (michalsnik) [inline tests/lib/rules/no-template-shadow.js]: Please add line on which the following error is expected

File: docs/rules/no-template-shadow.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
+# Disallow variable declarations from shadowing variables declared in the outer scope. (no-template-shadow)
+
+`no-shadow` should report variable definitions of v-for directives or scope attributes if those shadows the variables in parent scopes.
```
Comment (michalsnik) [inline docs/rules/no-template-shadow.md]: `no-template-shadow`

File: lib/utils/index.js (hunk 4 above)
Comment (michalsnik) [reply lib/utils/index.js]: It indicates an `optional param` http://usejsdoc.org/tags-param.html#optional-parameters-and-default-values

**Comment by michalsnik** [comment]
We're still in v5 beta. Let's put it in `strongly-recommended` so we can get feedback faster, and don't need to wait for the next major release to ship it by default.

**Pull request merged**